### DISCUSSION TITLE
Twitch/YouTube/etc chat integration for lazer

### DISCUSSION THREAD

**Discussion by willowrubenstein** [ppy/osu#15061] [Ideas] [answered]
Here's a brief bullet-point rundown as to how I think this could be implemented:

- Add it as a supporter feature
- Run Twitch/YT/Etc. IRC bot asynchronously with the game's runtime
  - This is added in options by pressing a button per each platform, which leads to a OAuth portal for each.
  - After a redirect, it will generate the codes per each chat
  - Ignore the first 2 steps if the platform doesn't require OAuth (I believe a few don't)
  - Check for messages in chat containing `https://osu.ppy.sh/beatmapsets`
  - Download through API hook into game, if the map doesn't already exist
- Allow customization options, such as BPM Max/Min, Star Max/Min, etc. through sliders in settings area when authorized.
  - These options will block maps that don't follow the options (will ignore map sets if there is a difficulty within the
defined min/max)
- Allow blacklisting users in the Streamer's chat for Moderators/Streamer, Whitelist
- Allow custom options like subs-only, followers-only, etc.
- Add cooldown per maps

These are just a few bullet points of ideas I had. I think that the ability to just keep playing the game without having to lookup/download maps and instead allow your chat to show you maps they liked would be huge. I'd love to see this start to 
get implemented. I could help myself if I had some pointers on what parts of lazer coorelate to where I'd need to work on this, 
as I'm not familiar with the mapping of the game's source.

Edit: should note that I am familiar with GitHub and this is a new account

**Comment by bdach** [comment]
almost all of this sounds like it is out of scope for the game. except maybe some sort of api that the irc bot could use to relay commands to the game. all of the rest (user voting, bans and whatnot) should not be a part of the game.

see also: https://github.com/ppy/osu/discussions/14945.

**Reply by willowrubenstein** [reply]
That would be neat. I've made my own build of this idea, but osu!api does not allow direct downloads at the current moment, so it has taken a lot of workarounds to make it work. However, implementing an IRC client into a Twitch chat actually honestly requires no authorization as long as you aren't chatting, if I remember correctly, so I'm not sure if it would be a HUGE hassle. I think that the first part is definitely possible.

**Reply by gagahpangeran** [reply]
You can relay twitch chat that contains beatmap link to your in game chat using mikuia bot for now.

And I agree all of the suggestion above are out of scope of this game.

**Comment by peppy** [comment, marked as answer]
I think the closest thing we have planned is some kind of api to interact with the game via websockets or otherwise. No actual integration like this is planned. That said, lazer is open source so nothing is stopping a competent third party from making it happen.

**Reply by willowrubenstein** [reply]
api would be sick!!!

**Reply by peppy** [reply]
See https://github.com/ppy/osu/pull/12274
